src/_pytest/pytester.py

```diff
@@ -359,8 +359,8 @@ def LineMatcher_fixture(request: FixtureRequest) -> "Type[LineMatcher]":
 
 
 @pytest.fixture
-def testdir(request: FixtureRequest, tmpdir_factory) -> "Testdir":
-    return Testdir(request, tmpdir_factory)
+def testdir(request: FixtureRequest, tmpdir_factory, monkeypatch) -> "Testdir":
+    return Testdir(request, tmpdir_factory, monkeypatch=monkeypatch)
```

Comment:
i noted that we missed a detail here,

we need to use `with monkeypatch.context() as mp:`

and pass that

finalize calls undo, which is a unintended sideeffect

Reply:
I've removed the call to undo.

Reply:
in terms of api, thats a breaking change

Reply:
You mean that previously calling `testdir.finalize()` would undo the changes to the environment?

This change here is about using the outer/global monkeypatch, so that you can override it from tests.
I've gone back to the simpler patch from my fork, with the removal of calling `self.monkeypatch.undo()`.

Reply:
> in terms of api, thats a breaking change

You mean the signature change in `Testdir.__init__`? If so, I disagree, `Testdir` is not part of the public API, only `testdir` is (the fixture). If someone is creating `Testdir` explicitly, it is doing so by importing from `_pytest.pytester`, which is clearly private API; that's why I believe it would be better to keep the previous version where `monkeypatch` is given as an argument.